pulp_rpm/app/models/repository.py

```diff
@@ -435,25 +435,21 @@ def _apply_retention_policy(self, new_version):
         ), "Cannot apply retention policy to completed repository versions"
 
         if self.retain_package_versions > 0:
-            # It would be more ideal if, instead of annotating with an age and filtering manually,
-            # we could use Django to filter the particular Package content we want to delete.
-            # Something like ".filter(F('age') > self.retain_package_versions)" would be better
-            # however this is not currently possible with Django. It would be possible with raw
-            # SQL but the repository version content membership subquery is currently
-            # django-managed and would be difficult to share.
-            #
-            # Instead we have to do the filtering manually.
-            nonmodular_packages = (
-                Package.objects.with_age()
-                .filter(
-                    pk__in=new_version.content.filter(pulp_type=Package.get_pulp_type()),
-                    is_modular=False,  # don't want to filter out modular RPMs
-                )
-                .only("pk")
-            )
+            # This code is likely not as efficient as it could / should be, but it is heavily
+            # complicated by the lack of a reliable means to determine whether a package is
+            # modular or not via SQL
+            packages = new_version.content.filter(pulp_type=Package.get_pulp_type())
+            modules = new_version.content.filter(pulp_type=Modulemd.get_pulp_type())
+
+            modular_packages = set()
+            for module in Modulemd.objects.filter(pk__in=modules):
+                modular_packages.update(module.packages.all().values_list("pk", flat=True))
+
+            nonmodular_packages = set(packages.values_list("pk", flat=True)) - modular_packages
+            nonmodular_packages = Package.objects.filter(pk__in=nonmodular_packages)
```

Comment:
@pedro-psb @ipanova Do either of you see a way to make this more efficient without table changes?

Making the `ModulemdPackage` table explicit rather than implicit would probably allow us to reduce the number of queries.  I'm unsure if it's worth it though.

Reply:
I did some research to see if I could come up with something but no luck.
There is `prefetch_related`, but I couldn't figure if this could fit in here.

Also, ideas of caching the context of the last repover crossed my mind, but I'm not aware of a suitable method for doing so.